go/types/map.go

```diff
@@ -38,34 +44,45 @@ func NewMap(kv ...Value) Map {
 	return newMap(ch.Done().(orderedSequence))
 }
 
-// NewStreamingMap takes an input channel of values and returns a output
-// channel that will produce a finished Map. Values sent to the input channel
-// must be alternating keys and values. (e.g. k1, v1, k2, v2...). Moreover keys
-// need to be added to the channel in Noms sortorder, adding key values to the
-// input channel out of order will result in a panic. Once the input channel is
-// closed by the caller, a finished Map will be sent to the output channel. See
-// graph_builder.go for building collections with values that are not in order.
-func NewStreamingMap(vrw ValueReadWriter, kvs <-chan Value) <-chan Map {
+// NewStreamingMap takes an input channel of values and returns an output
+// channel that will produce a StreamingMapResult. Values sent to the input
+// channel must be alternating keys and values. (e.g. k1, v1, k2, v2...).
+// Moreover keys must be non-nil and need to be added to the channel in Noms 
+// sortorder. If these constraints are violated, the result will contain a 
+// non-nil Error field and the map will only contain key-value pairs received on
+// the input channel before the first violation was found. Once the input channel
+// is closed by the caller, the StreamingMapResult will be sent to the output
+// channel. See graph_builder.go for building collections with values that are
+// not in order.
+func NewStreamingMap(vrw ValueReadWriter, kvs <-chan Value) <-chan StreamingMapResult {
```

Comment:
I don't think we should go to any trouble to allow callers to catch this condition gracefully. Just panic. If you want to test that case, then mock out the `go` statement or do something else to allow testing.